File: binary_search/binary_search.rb
def binary_search(list, target, imin = nil, imax = nil)
  imin ||= 0
  imax ||= list.length - 1
  if(imax < imin)
    return false
  else
    mid = imin + (imax - imin)/2
    if(list[mid] > target)
      return binary_search(list, target, imin, mid -1 )
    elsif(list[mid] < target)
      return binary_search(list, target, mid + 1, imax)
    else
      return mid
    end
  end
end


# ?? (imax + imin) / 2 = imin + (imax - imin) /2    ??
#
# (imax + imin ) /2 = x
# imax + imin = 2x
# imax - imin = 2x - 2imin
# (imax- imin) /2 = x - imin
# imin + (imax - imin) /2 = x
#
#
# (imax + imin) / 2  = imin + (imax - imin) /2
# imax + imin = 2imin + ( imax - imin )
# imax + imin = imax + imin

File: binary_search/binary_search.go
package main

import (
	"fmt"
	"sort"
)

func BinarySearch(list []int, target int, args ...int) int {
	imin := 0
	imax := len(list) -1

	if len(args) == 2 {
		imin = args[0]
		imax = args[1]
	}

	if imax < imin {
		return imax + 1
	} else {
		mid := imin + (imax-imin)/2
		if list[mid] > target {
			return BinarySearch(list, target, imin, mid-1)
		} else if list[mid] < target {
			return BinarySearch(list, target, mid+1, imax)
		} else {
			return mid
		}
	}
}

func main() {
	list := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15}
	fmt.Println(sort.SearchInts(list, 0))
	fmt.Println(BinarySearch(list, 5, 0, len(list)-1))
}
